build-tools/packages/build-cli/src/library/typescriptApi.ts

```diff
@@ -0,0 +1,97 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { SourceFile, ExportedDeclarations } from "ts-morph";
+import { Node } from "ts-morph";
+
+import { ApiLevel, isKnownApiLevel } from "./apiLevel.js";
+
+interface ExportRecord {
+	name: string;
+	isTypeOnly: boolean;
+}
+export interface ExportRecords {
+	public: ExportRecord[];
+	beta: ExportRecord[];
+	alpha: ExportRecord[];
+	internal: ExportRecord[];
+	unknown: { name: string; decl: ExportedDeclarations }[];
+}
+
+/**
+ * This function is a placeholder for functionality that could
+ * be very useful to have when working with multiple exports
+ * using the same name (basically both a type and value export).
+ */
+function isTypeExport(_decl: ExportedDeclarations): boolean {
+	return false;
+}
+
+/**
+ * Searches given Node's JSDocs for known {@link ApiLevel} tag.
+ *
+ * @returns Recognized {@link ApiLevel} from JSDocs or undefined.
+ */
+function getNodeLevel(node: Node): ApiLevel | undefined {
+	if (Node.isJSDocable(node)) {
+		for (const jsdoc of node.getJsDocs()) {
+			for (const tag of jsdoc.getTags()) {
+				const tagName = tag.getTagName();
+				if (isKnownApiLevel(tagName)) {
+					return tagName;
+				}
+			}
+		}
+	} else {
+		// Some nodes like `VariableDeclaration`s as not JSDocable, but an ancestor
+		// like `VariableStatement` is and may contain tag.
+		const parent = node.getParent();
+		if (parent !== undefined) {
+			return getNodeLevel(parent);
+		}
+	}
+	return undefined;
+}
+
+/**
+ * Given a source file extracts all of the named exports and associated API level.
```

Comment:
We should probably leave TODO notes in here for explicit namespaces and `export * as Foo` patterns. I know they aren't common, but they are supported by the language, so we should ideally support them here (or replace use existing tooling when it isn't broken 😋).

Also maybe hard fail on `export * as Foo` patterns if we don't support them?

Reply:
I think I have a handle on some more cases. Let's merge this and I'll follow with some updates. So far, I don't observe trouble from this logic with our current code.